### PR TITLE
ODT new generation

### PR DESCRIPTION
**Pull request by TeXitoi**
WiP ODT new generation

### CODE REVIEW

File: source/ed/connectors/fusio_parser.cpp [hunk 1]
```diff
@@ -275,6 +287,16 @@ void StopTimeFusioHandler::handle_line(Data& data, const csv_row& row, bool is_f
     }
 }
 
+template<typename T>
+static boost::optional<T> read_wkt(const std::string& s) {
+    try {
+        T g;
+        boost::geometry::read_wkt(s, g);
+        return g;
+    } catch (const boost::geometry::read_wkt_exception&) {
```
Comment (antoine-de) [inline source/ed/connectors/fusio_parser.cpp]: no log ?

File: source/ed/connectors/fusio_parser.cpp [hunk 2]
```diff
@@ -282,13 +304,13 @@ void GeometriesFusioHandler::init(Data&) {
...
-        nt::LineString line;
-        boost::geometry::read_wkt(row.at(geometry_wkt_c), line);
-        data.shapes[row.at(geometry_id_c)] = {line};
+        if (const auto g = read_wkt<nt::MultiLineString>(row.at(geometry_wkt_c))) {
+            data.shapes[row.at(geometry_id_c)] = *g;
+        } else if (const auto g = read_wkt<nt::LineString>(row.at(geometry_wkt_c))) {
+            data.shapes[row.at(geometry_id_c)] = {*g};
+        } else if (const auto g = read_wkt<nt::MultiPolygon>(row.at(geometry_wkt_c))) {
+            data.areas[row.at(geometry_id_c)] = *g;
+        }
```
Comment (antoine-de) [inline source/ed/connectors/fusio_parser.cpp:321]: I think you should add the log here

File: source/type/geographical_coord.h [hunk 3]
```diff
@@ -136,3 +139,25 @@ typedef std::vector<LineString> MultiLineString;
...
 BOOST_GEOMETRY_REGISTER_MULTI_LINESTRING(navitia::type::MultiLineString)
+
+namespace navitia { namespace type {
+    typedef boost::geometry::model::polygon<GeographicalCoord> Polygon;
+    typedef boost::geometry::model::multi_polygon<Polygon> MultiPolygon;
+}}
+namespace boost { namespace serialization {
+    template<class Archive>
+    void serialize(Archive& ar, navitia::type::Polygon& poly, const unsigned int) {
+        ar & boost::serialization::make_nvp("outer", poly.outer());
```
Comment (antoine-de) [inline source/type/geographical_coord.h:150]: you don't have to bother with the nvp, it's not handled anywhere :smile:

File: source/type/tests/multi_polygon_map_test.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE multi_polygon_map_test
+#include <boost/test/unit_test.hpp>
+
+#include "type/multi_polygon_map.h"
+
+using namespace navitia::type;
+
+BOOST_AUTO_TEST_CASE(inserting_diamon_and_find) {
```
Comment (antoine-de) [inline source/type/tests/multi_polygon_map_test.cpp]: I think you should add a bit more tests (with more than one shape for example)

File: source/type/tests/multi_polygon_map_test.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(6, 6));
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(5, 5));
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(4, 4));
+    BOOST_REQUIRE(!search.empty());
+    BOOST_CHECK_EQUAL(std::get<2>(*search.begin()), 42);
+
+    for (const auto& e: map) {}
```
Comment (antoine-de) [inline source/type/tests/multi_polygon_map_test.cpp]: debug

File: source/type/geographical_coord.h (hunk 3 above)
Comment (TeXitoi) [inline source/type/geographical_coord.h:150]: That's copy and past from the web ;)

File: source/ed/connectors/fusio_parser.cpp (hunk 1 above)
Comment (TeXitoi) [inline source/ed/connectors/fusio_parser.cpp]: logged elsewhere

File: source/type/multi_polygon_map.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        context_type ctx = {key, res};
+        const double coord[] = {key.lon(), key.lat()};
+        auto callback = [](const Value* val, void* ctx_ptr) -> bool {
+            context_type& ctx = *static_cast<context_type*>(ctx_ptr);
+            if (boost::geometry::within(ctx.first, val->first)) {
+                ctx.second.push_back(val->second);
+            }
+            return true;
+        };
+        rtree.Search(coord, coord, callback, &ctx);
```
Comment (antoine-de) [inline source/type/multi_polygon_map.h:76]: ouch :see_no_evil:

File: source/type/tests/multi_polygon_map_test.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    map.insert(elt, 42);
+    auto search = map.find(GeographicalCoord(7, 7));
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(6, 6));
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(5, 5));
+    BOOST_CHECK(search.empty());
+    search = map.find(GeographicalCoord(4, 4));
+    BOOST_REQUIRE_EQUAL(search.size(), 1);
+    BOOST_CHECK_EQUAL(search.at(0), 42);
```
Comment (antoine-de) [inline source/type/tests/multi_polygon_map_test.cpp:55]: can you add test with more elements  ?

File: source/type/multi_polygon_map.h (hunk 6 above)
Comment (TeXitoi) [inline source/type/multi_polygon_map.h:76]: Yeah, this r:deciduous_tree: :no_good:

File: source/type/tests/multi_polygon_map_test.cpp (hunk 7 above)
Comment (TeXitoi) [inline source/type/tests/multi_polygon_map_test.cpp:55]: Done with the Beatles

File: source/ed/connectors/fusio_parser.cpp [hunk 8]
```diff
@@ -36,9 +36,15 @@ namespace ed { namespace connectors {
 
 void AgencyFusioHandler::init(Data& data) {
     AgencyGtfsHandler::init(data);
+    if (id_c == -1) { id_c = csv.get_pos_col("network_id"); }
```
Comment (antoine-de) [inline source/ed/connectors/fusio_parser.cpp:39]: i think there is a is_valid(idx_c) helper somewhere

File: source/ed/connectors/fusio_parser.cpp [hunk 9]
```diff
@@ -1121,9 +1151,14 @@ ed::types::PhysicalMode* GtfsData::get_or_create_default_physical_mode(Data & da
 
 void FusioParser::parse_files(Data& data) {
     parse<GeometriesFusioHandler>(data, "geometries.txt");
-    parse<AgencyFusioHandler>(data, "agency.txt", true);
+
+    // TODO: we need at least one of agency.txt or networks.txt
```
Comment (antoine-de) [inline source/ed/connectors/fusio_parser.cpp]: I think this TODO is quite useful

File: source/jormungandr/jormungandr/interfaces/common.py [hunk 10]
```diff
@@ -29,11 +29,11 @@
 
 import navitiacommon.type_pb2 as type_pb2
 
-odt_levels = {"none","mixt", "zonal", "all"}
+odt_levels = {"scheduled","with_stops", "zonal", "all"}
 
 pb_odt_level = {
-    'none': type_pb2.odt_none,
-    'mixt': type_pb2.mixt,
+    'none': type_pb2.scheduled,
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/interfaces/common.py]: remove if useless, else update

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 11]
```diff
@@ -181,6 +181,15 @@ def output(self, key, obj):
         return str.lower(enum[getattr(obj, key)].name)
 
 
+class PbEnum(fields.Raw):
```
Comment (antoine-de) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py:184]: could you add a comment how to use it?

File: source/routing/raptor_api.cpp [hunk 12]
```diff
@@ -71,25 +72,46 @@ static void add_coord(const type::GeographicalCoord& coord, pbnavitia::Section*
...
 {
-    if (stop_times.empty()) return;
-    const auto& vj = stop_times.front()->vehicle_journey;
-    if (vj->is_odt() && vj->vehicle_journey_type != type::VehicleJourneyType::virtual_with_stop_time) {
-        for (const auto& st : stop_times) {
-            add_coord(st->journey_pattern_point->stop_point->coord, pb_section);
-        }
-        return;
-    }
+    if (stop_times.empty()) { return; }
```
Comment (antoine-de) [inline source/routing/raptor_api.cpp:75]: lash1 kom!

File: source/ed/connectors/fusio_parser.cpp (hunk 8 above)
Comment (TeXitoi) [inline source/ed/connectors/fusio_parser.cpp:39]: There isn't.

File: source/type/multi_polygon_map.h (hunk 6 above)
Comment (antoine-de) [inline source/type/multi_polygon_map.h:76]: so now there is the new debian stable you'll change it back to boost::geometry rtree ? :+1:

File: source/routing/raptor_api.cpp (hunk 12 above)
Comment (TeXitoi) [inline source/routing/raptor_api.cpp:75]: done

File: source/jormungandr/jormungandr/interfaces/v1/fields.py (hunk 11 above)
Comment (TeXitoi) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py:184]: done

**Pull request merged**